src/platform/datapath_raw_socket.c

```diff
@@ -320,14 +332,90 @@ CxPlatFramingWriteHeaders(
     _In_ const CXPLAT_SOCKET* Socket,
     _In_ const QUIC_ADDR* LocalAddress,
     _In_ const QUIC_ADDR* RemoteAddress,
-    _Inout_ QUIC_BUFFER* Buffer
+    _Inout_ QUIC_BUFFER* Buffer,
+    _In_ QUIC_ADDRESS_FAMILY Family
     )
 {
     UDP_HEADER* UDP = (UDP_HEADER*)(Buffer->Buffer - sizeof(UDP_HEADER));
-    IPV4_HEADER* IP = (IPV4_HEADER*)(((uint8_t*)UDP) - sizeof(IPV4_HEADER));
-    ETHERNET_HEADER* Ethernet = (ETHERNET_HEADER*)(((uint8_t*)IP) - sizeof(ETHERNET_HEADER));
+    ETHERNET_HEADER* Ethernet;
+    IPV4_HEADER* IPv4;
+    IPV6_HEADER* IPv6;
+    uint16_t EthType;
+    uint16_t IpHeaderLen;
+
+    CXPLAT_DBG_ASSERT(
+        Family == QUIC_ADDRESS_FAMILY_INET || Family == QUIC_ADDRESS_FAMILY_INET6);
+
+    //
+    // Fill UDP header.
+    //
+    UDP->DestinationPort = RemoteAddress->Ipv4.sin_port;
+    UDP->SourcePort = LocalAddress->Ipv4.sin_port;
+    UDP->Length = QuicNetByteSwapShort((uint16_t)Buffer->Length);
+
+    //
+    // Fill IPv4/IPv6 header.
+    //
+    if (Family == QUIC_ADDRESS_FAMILY_INET) {
+        IPv4 = (IPV4_HEADER*)(((uint8_t*)UDP) - sizeof(IPV4_HEADER));
+        IPv4->VersionAndHeaderLength = IPV4_DEFAULT_VERHLEN;
+        IPv4->TypeOfService = 0;
+        IPv4->TotalLength = htons(sizeof(IPV4_HEADER) + sizeof(UDP_HEADER) + (uint16_t)Buffer->Length);
+        IPv4->Identification = 0;
+        IPv4->FlagsAndFragmentOffset = 0;
+        IPv4->TimeToLive = IP_DEFAULT_HOP_LIMIT;
+        IPv4->Protocol = IPPROTO_UDP;
+        IPv4->HeaderChecksum = 0;
+        CxPlatCopyMemory(IPv4->Destination, &RemoteAddress->Ipv4.sin_addr, sizeof(RemoteAddress->Ipv4.sin_addr));
+        CxPlatCopyMemory(IPv4->Source, &LocalAddress->Ipv4.sin_addr, sizeof(LocalAddress->Ipv4.sin_addr));
+        EthType = ETHERNET_TYPE_IPV4;
+        Ethernet = (ETHERNET_HEADER*)(((uint8_t*)IPv4) - sizeof(ETHERNET_HEADER));
+        IpHeaderLen = sizeof(IPV4_HEADER);
+    } else {
+        IPv6 = (IPV6_HEADER*)(((uint8_t*)UDP) - sizeof(IPV6_HEADER));
+        //
+        // IPv6 Version, Traffic Class, ECN Field and Flow Label fields in host
+        // byte order.
+        // TODO: support big-endian systems. Not quite sure if it's needed since QUIC core code
+        // also doesn't support big-endian systems due to the use of CxPlatByteSwapXxx which swaps
+        // bytes even when there's no need.
```

Comment:
No need to have this explicit TODO here. As you've pointed out, all of MsQuic doesn't support big-endian already. We haven't had a need to do so yet.